pkg/deployment/cert.go

```diff
@@ -0,0 +1,163 @@
+				fqdnName := strings.Join([]string{nodeName, res.DNSDomain}, ".")
+				dnsNames = append(dnsNames, fqdnName)
+			}
+		}
+
+		// for now, just get a single cert with all the dnsnames for the host.  Its likely
+		// that this will be sufficient for most services.
+		// But it is clear how we can issue different certs as needed by calling IssueTLSCert
+		// and using ips instead of dnsnames, or by using a different issuer
+		secretName = "cert-" + nodeName
+		certName = secretName
+		certSecret, err := IssueTLSCert(ctx, helper, certName, instance.Namespace, secretName,
+			nodeName, dnsNames, nil, RootCAIssuerInternalNamespaceLabel,
+			RootCAIssuerInternalNamespaceLabel, labels)
+		if err != nil {
+			return err
+		}
+
+		// TODO(alee) Add an owner reference to the secret so it can be monitored
+		// We'll do this once stuggi adds a function to do this in libcommon
+
+		// To use this cert, add it to the relevant service data
+		for key, value := range certSecret.Data {
+			novaCertsData[nodeName+"-"+key] = value
+			libvirtCertsData[nodeName+"-"+key] = value
+		}
+	}
+
+	// create a secret for each service
+	novaCertsSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: instance.Name + "-nova-certs",
+		},
+		Data: novaCertsData,
+	}
+	secret.CreateOrPatchSecret(ctx, helper, instance, novaCertsSecret)
+
+	libvirtCertsSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: instance.Name + "-libvirt-certs",
+		},
+		Data: libvirtCertsData,
+	}
+	secret.CreateOrPatchSecret(ctx, helper, instance, libvirtCertsSecret)
+
+	return nil
+}
+
+func IssueTLSCert(ctx context.Context, helper *helper.Helper,
+	name string, namespace string, secretName string,
+	hostname string, dnsnames []string, ips []string,
+	issuerName string, issuerNamespace string,
+	labels map[string]string) (*corev1.Secret, error) {
+	issuer := &certmgrv1.Issuer{}
+	err := helper.GetClient().Get(ctx, types.NamespacedName{Name: issuerName, Namespace: issuerNamespace}, issuer)
+	if err != nil {
+		err = fmt.Errorf("Error getting issuer %s/%s - %w", issuerName, issuerNamespace, err)
+		return nil, err
+	}
+	certReq := certmanager.Cert(
+		name,
+		namespace,
+		labels,
+		certmgrv1.CertificateSpec{
+			CommonName:  hostname,
+			DNSNames:    dnsnames,
+			IPAddresses: ips,
+			Duration: &metav1.Duration{
+				Duration: time.Hour * 24 * 365,
+			},
+			IssuerRef: certmgrmetav1.ObjectReference{
+				Name:  issuer.Name,
+				Kind:  issuer.Kind,
+				Group: issuer.GroupVersionKind().Group,
+			},
+			SecretName: secretName,
+		},
+	)
+	// Note: 5 here is a timeout (in minutes?)
+	cert := certmanager.NewCertificate(certReq, 5)
+	_, err = cert.CreateOrPatch(ctx, helper)
+	if err != nil {
+		err = fmt.Errorf("Error issuing certificate %s - %w", name, err)
+		return nil, err
+	}
+
+	certSecret, _, err := secret.GetSecret(ctx, helper, secretName, namespace)
+	if err != nil {
+		err = fmt.Errorf("Error retrieving secret %s - %w", secretName, err)
```

Comment:
nit: there are two cases:
1) the Secret is being created by the cert-manager so we got NotFound here so we just need to wait with a RequeueAfter
2) we could not reach the API or the API failed in other reasons. A real error.

I would split the two cases as 1) is not an error for me

Reply:
ack

Reply:
looks resolved to me.